pkg/logging/log_viewer.go

```diff
@@ -0,0 +1,141 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package logging
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/sirupsen/logrus"
+)
+
+// Type alias for functions which write out logs to the provided stdout/stderr Writers.
+// Depending on the provided `LogViewingOptions.Follow` option, the function may block
+// indefinitely until something is sent through the `stopChannel`.
+type LogViewerFunc func(lvopts LogViewingOptions, stdout, stderr io.Writer, stopChannel chan os.Signal) error
+
+var logViewers = make(map[string]LogViewerFunc)
+
+// Registers a LogViewerFunc for the
+func RegisterLogViewer(driverName string, lvfn LogViewerFunc) {
+	if _, ok := logViewers[driverName]; ok {
+		logrus.Warnf("A LogViewerFunc with name %q has already been registered: %#v, overriding with %#v either way")
+	}
+	logViewers[driverName] = lvfn
+}
+
+func init() {
+	RegisterLogViewer("json-file", viewLogsJsonFile)
+	RegisterLogViewer("journald", viewLogsJournald)
+}
+
+// Returns a LogViewerFunc for the provided logging driver name.
+func getLogViewer(driverName string) (LogViewerFunc, error) {
+	lv, ok := logViewers[driverName]
+	if !ok {
+		return nil, fmt.Errorf("no log viewer type registered for logging driver %q", driverName)
+	}
+	return lv, nil
+}
+
+// Set of options passable to log viewers.
+type LogViewingOptions struct {
+	// Identifier (ID) of the container and namespace it's in.
+	ContainerId string
+	NamespaceId string
+
+	// Absolute path to the nerdctl datastore's root:
+	DatastoreRootPath string
+
+	// Whether or not to follow the output of the container logs.
+	Follow bool
+
+	// Whether or not to print timestampts for each line.
+	Timestamps bool
+
+	// Uint representing the number of most recent log entries to display. 0 = "all".
+	Tail uint
+
+	// Start/end timestampts to filter logs by.
+	Since string
+	Until string
+}
+
+func (lvo LogViewingOptions) InternalValidate(idsRequired bool) error {
+	if idsRequired && (lvo.ContainerId == "" || lvo.NamespaceId == "") {
+		return fmt.Errorf("log viewing options require a ContainerId and NamespaceId: %#v", lvo)
+	}
```

Comment:
we always need `ContainerId` and `Namespace`. This check is useless

Reply:
Do you mean the arg `isRequred` was useless for this function (which I have removed as I agree), or that we should not allow `LogViewOptions` to ever be created without a ContainerID/Namespace? (in which case I'd private its fields and make a function to build them which would perform this check)

Reply:
I mean the arg `isRequired` was useless for this function 👍